### PR TITLE
Sheep compatiability

### PR DESCRIPTION
**Pull request by shardros**
This PR will be merged when there is all of the parts which sheep needs to run working.

### CODE REVIEW

**Comment by fenjalien** [comment]
Running this branch on my pi produces this error when the start button is pressed in the editor: 
```
INFO:     127.0.0.1:44546 - "POST /upload/upload HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 394, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 580, in __call__
    await route.handle(scope, receive, send)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 241, in handle
    await self.app(scope, receive, send)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 52, in app
    response = await func(request)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/fastapi/routing.py", line 216, in app
    raw_response = await run_endpoint_function(
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/fastapi/routing.py", line 151, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/home/pi/.cache/pypoetry/virtualenvs/shepherd-2-TGIJcPDa-py3.10/lib/python3.10/site-packages/starlette/concurrency.py", line 40, in run_in_threadpool
    return await loop.run_in_executor(None, func, *args)
  File "/home/pi/.pyenv/versions/3.10.6/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/pi/shepherd-2/./app/routers.py", line 55, in upload_file
    app.upload.process_uploaded_file(uploaded_file)
  File "/home/pi/shepherd-2/./app/upload.py", line 110, in process_uploaded_file
    shutil.rmtree(config.round_path)
  File "/home/pi/.pyenv/versions/3.10.6/lib/python3.10/shutil.py", line 714, in rmtree
    onerror(os.lstat, path, sys.exc_info())
  File "/home/pi/.pyenv/versions/3.10.6/lib/python3.10/shutil.py", line 712, in rmtree
    orig_st = os.lstat(path)
FileNotFoundError: [Errno 2] No such file or directory: '/home/pi/shepherd-2/usercode/round'
```
The later run request gives a 409
```info
INFO:     127.0.0.1:49338 - "POST /run/start HTTP/1.1" 409 Conflict
```

The output of running pytest is 
```
test/test_editor.py ..                                                             [ 20%]
test/test_home.py .                                                                [ 30%]
test/test_runner.py x..F                                                           [ 70%]
test/test_upload.py FFF                                                            [100%]
```
It looks like the failures are also due to the directory `usercode/round` not exsisting, what is the purpose of this folder @shardros ?

**Comment by shardros** [comment]
Fixed and tested on fresh install

**Comment by shardros** [comment]
> It looks like the failures are also due to the directory usercode/round not exsisting, what is the purpose of this folder @shardros ?

This is correct. This is the folder where the usercode which is to be run in the next round is uploaded to. It is now automatically created by shepherd. Logs aren't done in this pr but you can see that the code works by looking `logs.txt`

**Comment by fenjalien** [comment]
> This is the folder where the usercode which is to be run in the next round is uploaded to.

Ah so `usercode/` is all the code that has been written by the user and `usercode/round/` is the code that will be run when `/run/start` is hit?

**Comment by shardros** [comment]
> > This is the folder where the usercode which is to be run in the next round is uploaded to.
> 
> Ah so `usercode/` is all the code that has been written by the user and `usercode/round/` is the code that will be run when `/run/start` is hit?

Correct :)

**Review by fenjalien — CHANGES_REQUESTED** [review]
From running the tests: 
```
test/test_editor.py ..        [ 20%]
test/test_home.py .         [ 30%]
test/test_runner.py xF..   [ 70%]
test/test_upload.py .FF    [100%]
```
`test_start` fails at line [20](https://github.com/systemetric/shepherd-2/blob/18c9b5400693804166bce94a13cfecd273b374ea/test/test_runner.py#L20) due to a timeout. Looks like it comes from [this line](https://github.com/systemetric/shepherd-2/blob/18c9b5400693804166bce94a13cfecd273b374ea/app/routers.py#L33):
```python
type(runner).state.__set__(runner, States.RUNNING)
```
Why is it like this? In fact all of the rest of the tests fail due to this.

**Comment by fenjalien** [comment]
I've just tried changing this: https://github.com/systemetric/shepherd-2/blob/18c9b5400693804166bce94a13cfecd273b374ea/app/routers.py#L33
to this:
```python
runner.state = States.RUNNING
```
and all tests pass apart from `larg_zip_upload` by timeout. I'm assuming this is due to the shear size its timeout will need to be much longer.

File: app/editor.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+def delete_file(filename):
+    if filename == "blocks.json":
+        return ""
+    dots = len(re.findall("\\.", filename))
+    if dots == 1:
+        os.unlink(os.path.join(config.usr_src_path, filename))
+    else:
+        _logger.warn("A file was attempted to be saved with too many dots: "
+                     f"{filename}")
```
[the remark this answers is not in the document]
Comment (shardros) [reply app/editor.py:69]: > From running the tests:
> 
> ```
> test/test_editor.py ..        [ 20%]
> test/test_home.py .         [ 30%]
> test/test_runner.py xF..   [ 70%]
> test/test_upload.py .FF    [100%]
> ```
> 
> `test_start` fails at line [20](https://github.com/systemetric/shepherd-2/blob/18c9b5400693804166bce94a13cfecd273b374ea/test/test_runner.py#L20) due to a timeout. Looks like it comes from [this line](https://github.com/systemetric/shepherd-2/blob/18c9b5400693804166bce94a13cfecd273b374ea/app/routers.py#L33):
> 
> ```python
> type(runner).state.__set__(runner, States.RUNNING)
> ```
> 
> Why is it like this? In fact all of the rest of the tests fail due to this.

This was made in error, I have fixed this and just pushed this.

I've added some doc-strings. I'm not entirely clear on why they need to have the dots removed this is code coppied from `shepherd-1`

File: app/config.py [hunk 2]
```diff
@@ -20,7 +21,58 @@ class Settings(BaseSettings):
...
+    def __init__(self):
+        self._init_usercode()
+
+    def _get_team_specifics(self):
+        """Find infomation set on each brain about the team"""
+        # Teamname should be set on a per brain basis before shipping
+        # Its purpose is to allow the setting of specific graphics for help identifing teams in the arena.
+        # Graphics are loaded from the ArenaUSB stick if available, or standard graphics from the stick are used.
+        # this used to be in rc.local, but the looks of shame and dissapointment
+        # got the better of me
```
[the remark this answers is not in the document]
Comment (shardros) [reply app/config.py]: Re-aranged to be in the docstring

File: app/config.py [hunk 3]
```diff
@@ -20,7 +21,58 @@ class Settings(BaseSettings):
...
+        if teamname_file.exists():
+            teamname_jpg = teamname_file.read_text().replace('\n', '') + '.jpg'
+        else:
+            teamname_jpg = 'none'
+
+        # Pick a start imapge in order of preference :
+        #     1) We have a team corner image on the USB
+        #     2) The team have uploaded their own image to the robot
+        #     3) We have a generic corner image on the USB
+        #     4) The game image
```
[the remark this answers is not in the document]
Comment (shardros) [reply app/config.py]: Resolved into doc string

File: app/config.py [hunk 4]
```diff
@@ -20,7 +21,58 @@ class Settings(BaseSettings):
...
+            # attempt to find the team specific corner graphic from the ArenaUSB
+            start_graphic = Path('robotsrc/team_logo.jpg')
+        if not start_graphic.exists():
+            # attempt to find the default corner graphic from ArenaUSB
+            start_graphic = game_control_path / 'Corner.jpg'
+        if not start_graphic.exists():
+            # finally look for a game specific logo
+            start_graphic = Path('/home/pi/game_logo.jpg')
+
+        if config.start_graphic.exists():
```
[the remark this answers is not in the document]
Comment (shardros) [reply app/config.py]: Start graphic was set unconditionally on line 48

File: app/routers.py [hunk 5]
```diff
@@ -1,23 +1,27 @@
...
+from pydantic import BaseModel
 
-from app.runner import States
-from app.runner import runner
+from app.run import States
+from app.run import runner
+import app.editor
 import app.upload
 
+_logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (shardros) [reply app/routers.py:10]: FastAPI has a really not-pythonic logging interfaces I have switched to using `loguru` which does logging much better in my opion and the logs should now appear everywhere.

File: app/routers.py [hunk 6]
```diff
@@ -43,13 +47,36 @@ def output():
...
+    contents: str
+
+
+@files_router.get('/')
+def get_files():
+    return app.editor.get_files()
+
+
+@files_router.post("/save/{filename}")
+async def save_file(filename: str, request: Request):
```
[the remark this answers is not in the document]
Comment (shardros) [reply app/routers.py:75]: `request.body()` requires to be awaited on which has to happen inside of an `async` function and I didn't convert all of the others to async. We probally should do that at some point but it's not on the critical path to reach our MVP as shepherd-1 does not do this.
https://github.com/systemetric/shepherd-2/issues/15

File: app/routers.py [hunk 7]
```diff
@@ -43,13 +47,36 @@ def output():
...
     }
 
+# ==============================================================================
+# Files router
+# ==============================================================================
+
+files_router = APIRouter(prefix="/files")
+
+class SheepFile(BaseModel):
+    contents: str
```
[the remark this answers is not in the document]
Comment (shardros) [reply app/routers.py:66]: Not sure what you mean by "singular" would you be able to explain?

File: install.sh [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    exit
+fi
+
+echo "-------------------------------------------------------------------------------------"
+echo "Poetry failed to install dependancies. Maybe there is no good python version."
+echo "You can manually do that here:"
+echo "  - Guide: https://realpython.com/installing-python/#how-to-build-python-from-source-code"
+echo "  - Offical python docs: https://docs.python.org/3/using/unix.html"
+echo "-------------------------------------------------------------------------------------"
+echo "SOLUTION: install.sh is able to compile python3.10 from source do you want to try?"
```
[the remark this answers is not in the document]
Comment (shardros) [reply install.sh:38]: Agreed, I think it is worth it never the less.

We later check about swap size:
```
# This is a problem on Pi's
echo "Please confirm that you are on a system with >2GB of memory+swap (increase swap if not) [y/n] "
old_stty_cfg=$(stty -g)
stty raw -echo
answer=$( while ! head -c 1 | grep -i '[ny]' ;do true ;done )
stty $old_stty_cfg
if echo "$answer" | grep -iq "^n" ;then
    exit
fi
```

File: install.sh [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    echo "INSTALL COMPLETE: Use 'sh run.sh' to run"
+    echo "========================================"
+    exit
+fi
+
+echo "-------------------------------------------------------------------------------------"
+echo "Poetry failed to install dependancies. Maybe there is no good python version."
+echo "You can manually do that here:"
+echo "  - Guide: https://realpython.com/installing-python/#how-to-build-python-from-source-code"
+echo "  - Offical python docs: https://docs.python.org/3/using/unix.html"
```
[the remark this answers is not in the document]
Comment (shardros) [reply install.sh:36]: Done in commit.

File: app/routers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shardros) [reply app/routers.py:75]: Let me know if you disagree with me on this being the right decision

**Comment by shardros** [comment]
@fenjalien ready for another round of review

**Comment by shardros** [comment]
You can still retroactively do the code review but we need to move fast

**Pull request merged**